File: parser/src/main/java/de/dst/fortran/code/generator/java/UnitGenerator.java
package de.dst.fortran.code.generator.java;

import com.helger.jcodemodel.*;
import de.dst.fortran.code.*;
import de.irt.jfor.Unit;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

import java.util.Objects;

import static de.dst.fortran.code.Analyzer.childElement;
import static de.dst.fortran.code.Analyzer.childElements;

/**
 * Created by IntelliJ IDEA.
 * User: stueken
 * Date: 21.10.17
 * Time: 15:35
 */
class UnitGenerator extends MethodGenerator {

    final CodeElement element;

    UnitGenerator(CodeGenerator codeGenerator, CodeElement element, JDefinedClass jclass) {
        super(codeGenerator, jclass);
        this.jclass._extends(de.irt.jfor.Unit.class);
        this.element = element;
    }

    void define() {
        try {
            constructor();
            decl();
            func();
            body();
        } catch(Throwable e) {
            String message = String.format("error parsing %s.%s:%s", element.code().path, element.code().name, line);
            throw new RuntimeException(message, e);
        }
    }
    
    void constructor() {
        JMethod constructor = jclass.constructor(JMod.PUBLIC);
        JVar parent = constructor.param(JMod.NONE, Unit.class, "parent");
        constructor.body().add(JExpr.invokeSuper().arg(parent));
    }

    void decl() {

        JDocComment comment = null;

        for (Common common : element.code().commons) {
            JDefinedClass jcommon = codeGenerator.commons.get(common.getName());
            JFieldVar cvar = jclass.field(JMod.PRIVATE | JMod.FINAL,
                    jcommon, common.getName(),
                    JExpr.invoke("common").arg(JExpr.dotclass(jcommon)));
            if (comment == null)
                comment = cvar.javadoc();

            // define variables for each member using local names
            common.members().forEach(member->{
                JFieldVar var = jcommon.fields().get(member.getRefName());
                JFieldRef ref = cvar.ref(var);
                decl(ref, member.getName());
            });
        }

        if (comment != null) {
            comment.setSingleLineMode(true);
            comment.add("\n");
            comment.add("commons");
            comment = null;
        }

        for (Code block : element.code().blocks) {
            String name = block.name;
            JDefinedClass junit = codeGenerator.units.get(name).jclass;
            JFieldVar jvar = jclass.field(JMod.PRIVATE | JMod.FINAL,
                    junit, name,
                    JExpr.invoke("unit").arg(JExpr.dotclass(junit)));
            if (comment == null)
                comment = jvar.javadoc();
            decl(jvar);
        }

        if (comment != null) {
            comment.setSingleLineMode(true);
            comment.add("\n");
            comment.add("units");
            comment = null;
        }

        for (Variable var : element.code().variables) {
            if (var.context == null && !var.isPrimitive()) {
                JFieldVar jvar = codeGenerator.defineVariable(jclass, var, JMod.PRIVATE);
                if (comment == null)
                    comment = jvar.javadoc();
                decl(jvar);
            }
        }

        if (comment != null) {
            comment.setSingleLineMode(true);
            comment.add("\n");
            comment.add("variables");
            comment = null;
        }
    }

    void func() {
        Element functions = childElement(element.getElement(), "functions");
        childElements(functions, "function").forEach(this::function);
    }

    void function(Element fun) {
        String name = fun.getAttribute("name");
        Variable var = element.code().variables.find(name);
        if(var==null || var.context!= Context.FUNCTION)
            throw new IllegalArgumentException("undefined function: " + name);

        Element assarr = childElement(fun, "assarr");

        MethodGenerator method = new MethodGenerator(codeGenerator, jclass);
        method.method(JMod.PUBLIC|JMod.STATIC, var.type(), name);
        method.jmethod.mods();

        Entities<Variable> arguments = new Entities<>(Variable::new);

        childElements(childElement(assarr, "args"), "arg").stream()
                .flatMap(arg -> childElements(arg, "var").stream())
                .map(v->v.getAttribute("name"))
                .map(arguments::get)
                .forEach(method::param);

        JDocComment comment = method.jmethod.javadoc();
        for (Element ce : childElements(fun)) {
            switch (ce.getTagName()) {
                case "F":
                    line = ce.getAttribute("line");
                    break;

                case "C":
                    String text = ce.getTextContent();
                    comment.add(text);
                    comment.add("\n");
                    break;
            }
        }

        method.line = line;
        final JBlock jbody = method.jmethod.body();

        // single line expression
        jbody._return(method.expr(childElements(assarr, "expr")));
    }

    void body() {

        method(JMod.PUBLIC, element.code().type(), "call");

        header(childElement(element.getElement(), "decl"));

        // prepare arguments
        for (Element arg : childElements(childElement(element.getElement(), "args"), "arg")) {

            boolean nl = false;
            for (Element ce : childElements(arg)) {
                switch (ce.getTagName()) {
                    case "F":
                        line = ce.getAttribute("line");
                        break;
                    case "f":
                        nl = true;
                        break;
                    case "var": {
                        Variable var = element.code().arguments.get(ce.getAttribute("name"));
                        JVar param = param(var);
                        if (nl)
                            param.annotations();
                        break;
                    }
                }
            }
        }

        //int n = 0;
        //for (Variable arg : code.block().arguments) {
        //    ++n;
        //    final JVar param = jmethod.param(JMod.FINAL, arg.type(), arg.name);
        //    if ((n % 6) == 5) {
        //        param.annotations();
        //        //param.annotate(Comment.class);
        //        //decl(param);
        //    }
        //}

        final JBlock jbody = jmethod.body();
        jbody.add(JFExpression.NL);

        // local variables
        for (Variable var : element.code().variables) {
            if (var.context == null && var.isPrimitive()) {

                final Class<?> type = codeGenerator.typeOf(var.type());
                AbstractJType jType = codeGenerator.codeModel._ref(type);
                IJExpression init = Boolean.class.isAssignableFrom(type) ? JExpr.FALSE : JExpr.lit(0);
                JVar jvar = jbody.decl(jType, var.name, init);
                decl(jvar);
            }
        }

        if (!element.code().variables.isEmpty())
            jbody.add(JFExpression.NL);

        childElements(childElement(element.getElement(), "code"))
                .stream()
                .map(this::code)
                .filter(Objects::nonNull)
                .forEach(jbody::add);
    }

    void header(Element decl) {
        if (decl == null)
            return;

        JDocComment comment = jmethod.javadoc();

        for (Element ce : childElements(decl)) {
            switch (ce.getTagName()) {
                case "F":
                    line = ce.getAttribute("line");

                case "C":
                case "f":
                    comment.add(ce.getTextContent());
                    comment.add("\n");
                    break;
            }
        }

        for (Node node = decl.getFirstChild(); node != null; node = node.getNextSibling()) {

            if (node.getNodeType() == Node.COMMENT_NODE || "C".equals(node.getNodeName())) {
                comment.add(node.getTextContent());
                comment.add("\n");
            }
        }
    }
}

File: parser/src/main/java/de/dst/fortran/code/generator/kotlin/UnitGenerator.kt
package de.dst.fortran.code.generator.kotlin

import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.TypeSpec
import de.dst.fortran.code.CodeElement
import de.dst.fortran.code.TypeDef
import de.dst.fortran.code.Variable
import de.irt.kfor.Fortran
import de.irt.kfor.Units
import org.w3c.dom.Element
import kotlin.reflect.KClass

/**
 * Created by IntelliJ IDEA.
 * User: stueken
 * Date: 13.01.19
 * Time: 18:17
 */

fun CodeElement.camelName(): String {

    var name = this.code().name;

    var c = name[0];
    if (c.isUpperCase())
        return name;

    return c.toUpperCase() + name.substring(1)
}

class UnitGenerator(generators : CodeGenerators, override val block : CodeElement, className : ClassName)
    : ClassGenerator(generators, "function", className), Generator {

    companion object {
        fun create(generators: CodeGenerators, element: CodeElement): UnitGenerator {
            val className = ClassName(generators.packageRoot + '.' + element.code().path, element.camelName())
            return UnitGenerator(generators, element, className)
        }

        fun blocks(generators: CodeGenerators): CodeBlocks<CodeElement> {
            return object : CodeBlocks<CodeElement>(generators) {
                override fun generate(block: CodeElement) = UnitGenerator.create(generators, block);
            }
        }
    }
    override val generator = this

    var lineNumber : Int = block.line.toInt()

    fun setLineNumber(line : String?) {
        lineNumber = if(line==null || line.isEmpty()) 0 else line.toInt()
        debug(block.name, lineNumber)
    }

    fun debug(name : String, line : Int) {

    }

    override fun getKlass(type : TypeDef?) : KClass<*> = generators.types.get(type)

    val type : KClass<*> = getKlass(code.getReturnType().primitive())

    override fun targetName(variable : Variable, asReference: Boolean) : String {
        var target = ""

        if(variable.context!=null) {
            if(variable.context!=code)
                target = "${variable.context.name}."
        }

        target += "%N"

        if(!asReference && variable.isProperty())
            target += ".v"

        return target
    }

    // variable
    override val retval = if(Unit::class==type) null else
        Variable(block.name)
                .type(code.getReturnType())
                .prop(Variable.Prop.ASSIGNED)
                .prop(Variable.Prop.RETURNED)

    val functions = mutableMapOf<String, LocalFunction>()


    override fun getVariable(name: String) : Variable {

        if(name == block.name)
            return retval!!

        val v = code.variables.find(name)!!

        if(v.ref==null)
            return v

        if(v.isIndex())
            return v.ref

        if(!v.ref.isIndex())
            return v.ref;

        return v;
    }

    fun Variable.isMember(): Boolean = context == null && !isPrimitive() && name != code.name

    override fun toString() : String = "${code.path}.${code.name}:${lineNumber}"

    override fun generate() {
        try {
            super.generate();
        } catch (error: Throwable) {
            throw RuntimeException("error building ${this}", error)
        }
    }

    override fun TypeSpec.Builder.generate(): TypeSpec.Builder {

        val name = block.element.name

        block.element["decl"].all("C")
                .forEach{e -> addKdoc(e.textContent.replace("%", "%%") + "\n")}

        superclass(Fortran::class)

        primaryConstructor(FunSpec.constructorBuilder()
                .addParameter("units", Units::class)
                .build())

        addSuperclassConstructorParameter(CodeBlock.of("units"))

        for (common in code.commons) {
            addProperty(generators.asProperty(common))
        }

        for (unit in code.blocks) {
            addProperty(generators.asProperty(unit))
        }

        for (variable in code.variables) {
            if (variable.isMember())
                addProperty(variable.asProperty())
        }

        for (func in block.element["functions"].all("function")) {
            addFunction(localFunction(func))
        }

        val datas = block.element["decl"].all("data")

        if(datas.isNotEmpty()) {
            val initializers = CodeBlock.builder()
            for(data in datas) {
                initializers.add(dataBlock(data))
            }
            addInitializerBlock(initializers.build())
        }
        
        addFunction(mainFunction())

        return this;
    }

    private fun dataBlock(el: Element) : CodeBlock {

        val code = object : CodeBuilder(this) {
            init {
                val names = el.all("var")
                var values : MutableList<Element> = el["values"].children()

                for(name in names) {
                    val variable = getVariable(name.name)

                    when(variable.dim.size) {
                        0 -> initVar(variable, values)
                        1 -> initArr(variable, values)
                        2 -> initMat(variable, values)
                    }

                    code.add("\n")
                }

                if(values.isNotEmpty())
                    throw RuntimeException("remaining data values")
            }

            private fun initVar(variable: Variable, values: MutableList<Element>) {
                val target = targetName(variable, false)
                code.add("«$target = ", variable.name)
                code.add(getValues(values, 1))
                code.add("»")
                endLine(values)
            }

            private fun initArr(variable: Variable, values: MutableList<Element>) {
                val target = targetName(variable, false)
                code.add("«$target.assign(\n", variable.name)
                val dim = variable.dim[0].toInt()
                code.add(getValues(values, dim))
                code.add(")»")
                endLine(values)
            }

            private fun initMat(variable: Variable, values: MutableList<Element>) {
                val target = targetName(variable, false)

                val nj = variable.dim[1].toInt()
                val ni = variable.dim[0].toInt()

                for(i in 1 .. nj) {
                    code.add("«$target($i).assign(\n", variable.name)
                    code.add(getValues(values, ni))
                    code.add(")»")
                    endLine(values)
                }
            }

            private fun getValues(values: MutableList<Element>, dim : Int) : CodeBlock {
                return expr().addValues(values, dim).build()
            }

            fun endLine(values: MutableList<Element>) {

                // pending comments etc.
                v@while(values.isNotEmpty()) {
                    val elem = values.get(0)
                    when(elem.getTagName()) {
                        "F" -> codeLine(elem)
                        "f"  -> contLine(elem)
                        "c" -> comment(elem)
                        "C" -> commentLine(elem)
                        else -> {
                            break@v
                        }
                    }
                    values.removeAt(0)
                }
            }
        }

        return code.build();
    }

    fun localFunction(func: Element) : FunSpec {
        val lf = LocalFunction(this, func)
        functions.put(lf.element.name, lf)
        return lf.build()
    }

    fun mainFunction() = MainFunction(this).build()
}

